### PR TITLE
feat(textarea): add label slot

### PR DESCRIPTION
**Pull request by liamdebeasi**
Issue number: N/A

---------

## What is the current behavior?

Textarea only accepts plaintext labels.

## What is the new behavior?

- Textarea can now accept custom HTML labels via a slot

A few implementation notes:

- Support for dynamically adding/removing slotted content will be added in a separate PR.
- Support for the notch outline will be added in a separate PR.
- The existing start/end placement label overflow tests were made redundant with the new tests I added. As a result, I removed the old tests.

## Does this introduce a breaking change?

- [ ] Yes
- [x] No

### CODE REVIEW

File: core/src/components/textarea/textarea.scss [hunk 1]
```diff
@@ -67,13 +67,27 @@
...
+   * textarea defaults to break-spaces which
+   * can cause extra whitespace to appear
+   * when using newlines before a slotted
+   * label element. To avoid this, we use
+   * "collapse" which is the default value
+   * for an <input />
+   * If textarea is converted to Shadow DOM
+   * then this may not be needed.
+   */
+  white-space-collapse: collapse;
```
Comment (liamdebeasi) [inline core/src/components/textarea/textarea.scss:81]: `<input />` defaults to `collapse` whereas textarea defaults to `preserve` which is why this change was only needed on ion-textarea.

File: core/src/components/textarea/textarea.scss [hunk 2]
```diff
@@ -74,6 +72,8 @@
   flex: 1;
 
   background: var(--background);
+
+  white-space: pre-wrap;
```
Comment (liamdebeasi) [inline core/src/components/textarea/textarea.scss:76]: This change is necessary because the old code was causing whitespace to appear in the textarea if you added newlines when slotted in a label:

```html
<ion-textarea label-placement="start" value="hi@ionic.io">
  <div slot="label">Email <span class="required">*</span></div>
</ion-textarea>
```

This code was added as a result of https://github.com/ionic-team/ionic-framework/issues/18495 to make it so you can set `white-space` on `ion-textarea` and have it inherit to the native textarea. This is not a pattern we should recommend using the modern form controls as it will impact both the textarea and the label, which is not always what developers want. Instead we should encourage developers to target the label or the textarea directly when changing white space.

Comment (liamdebeasi) [reply core/src/components/textarea/textarea.scss:76]: You'll also see a few screenshot diffs because this is impacting textareas in `main`: https://github.com/ionic-team/ionic-framework/pull/27647/commits/cf85fa4a4164d96d3408cdb7f27488ff5d773cf2. The buggy behavior is not common right now but it will likely become more common once label slots become available.

File: core/src/components/textarea/textarea.tsx [hunk 3]
```diff
@@ -530,17 +536,37 @@ Developers can use the "legacy" property to continue using the legacy form marku
...
     return (
-      <div class="label-text-wrapper">
-        <div class="label-text">{this.label}</div>
+      <div
+        class={{
+          'label-text-wrapper': true,
+          'label-text-wrapper-hidden': !this.hasLabel,
+        }}
+      >
+        {label === undefined ? <slot name="label"></slot> : <div class="label-text">{label}</div>}
```
Comment (thetaPC) [inline core/src/components/textarea/textarea.tsx:547]: The screen reader (Voice Over) is reading the label three time when the textarea is using a slot and twice for when it's using a label prop. This isn't related to the ticket but it would be good to create a spike ticket for further investigation.

**Review by thetaPC — APPROVED** [review]
LGTM

File: core/src/components/textarea/textarea.tsx (hunk 3 above)
Comment (liamdebeasi) [reply core/src/components/textarea/textarea.tsx:547]: How are you testing this? I tested this with VoiceOver on macOS (Safari and Chrome) and iOS 16+17 and the screen reader announces the slotted label once.

Comment (liamdebeasi) [reply core/src/components/textarea/textarea.tsx:547]: Ah I can reproduce this on Firefox. However, I can reproduce this outside of Ionic too so I think this may be a platform bug (either with Firefox or VoiceOver). The following reads "Hello World, edit text, Hello World, group":

```html
<label>
  Hello World
  <textarea></textarea>
</label>
```

However, the following reads "Hello World, edit text, main":

```html
<label for="text">Hello World</label>
<textarea id="text"></textarea>
```

Comment (liamdebeasi) [reply core/src/components/textarea/textarea.tsx:547]: Filed a bug: https://bugzilla.mozilla.org/show_bug.cgi?id=1838502

Comment (thetaPC) [reply core/src/components/textarea/textarea.tsx:547]: Interesting indeed. I agree that it's an external bug. Then I vote to create a spike to figure out where and submit a bug ticket to them if there isn't one.

Comment (thetaPC) [reply core/src/components/textarea/textarea.tsx:547]: I can verify that this is also happening with `ion-radio`. Works great for Chrome but not Firefox.

File: core/src/components/textarea/test/label-placement/textarea.e2e.ts [hunk 4]
```diff
@@ -234,3 +211,32 @@ configs().forEach(({ title, screenshot, config }) => {
     });
   });
 });
+
+configs({ modes: ['md'], directions: ['ltr'] }).forEach(({ title, screenshot, config }) => {
+  test.describe(title('textarea: label overflow'), () => {
+    test('label property should be truncated with ellipses', async ({ page }) => {
```
Comment (brandyscarney) [inline core/src/components/textarea/test/label-placement/textarea.e2e.ts:217]:
```suggestion
    test('label property should be truncated with an ellipsis', async ({ page }) => {
```

Nitpick, feel free to ignore but singular is ellipsis

File: core/src/components/textarea/test/label-placement/textarea.e2e.ts [hunk 5]
```diff
@@ -234,3 +211,32 @@ configs().forEach(({ title, screenshot, config }) => {
...
+        `
+            <ion-textarea label="Label Label Label Label Label" placeholder="Text Input"></ion-textarea>
+          `,
+        config
+      );
+
+      const textarea = page.locator('ion-textarea');
+      expect(await textarea.screenshot()).toMatchSnapshot(screenshot(`textarea-label-truncate`));
+    });
+    test('label slot should be truncated with ellipses', async ({ page }) => {
```
Comment (brandyscarney) [inline core/src/components/textarea/test/label-placement/textarea.e2e.ts:228]:
```suggestion
    test('label slot should be truncated with an ellipsis', async ({ page }) => {
```

Nitpick, feel free to ignore but singular is ellipsis

File: core/src/components.d.ts [hunk 6]
```diff
@@ -2921,6 +2921,9 @@ export namespace Components {
          */
         "mode"?: "ios" | "md";
     }
+    /**
+     * @experimental The label text to associate with the textarea. Use the `labelPlacement` property to control where the label is placed relative to the textarea. Use this if you need to render a label with custom HTML.
```
Comment (brandyscarney) [inline core/src/components.d.ts:2925]: Why is this experimental tag showing up on top of textarea but it doesn't mention the actual `label` slot or property?

File: core/src/components/textarea/textarea.scss (hunk 2 above)
Comment (brandyscarney) [reply core/src/components/textarea/textarea.scss:76]: This makes sense to me.

File: core/src/components/textarea/textarea.tsx [hunk 7]
```diff
@@ -15,6 +15,8 @@ import type { TextareaChangeEventDetail, TextareaInputEventDetail } from './text
 
 /**
  * @virtualProp {"ios" | "md"} mode - The mode determines which platform styles to use.
+ *
+ * @slot label - @experimental The label text to associate with the textarea. Use the `labelPlacement` property to control where the label is placed relative to the textarea. Use this if you need to render a label with custom HTML.
```
Comment (brandyscarney) [inline core/src/components/textarea/textarea.tsx:19]: Is `@experimental` in the right place here? It shows up weird in the component file

**Review by brandyscarney — APPROVED** [review]
Looks good besides the notch outline calculation which will be done in a separate PR. 👍

File: core/src/components.d.ts (hunk 6 above)
Comment (liamdebeasi) [reply core/src/components.d.ts:2925]: Not sure. Looks like this is also happening for the input slot too. Let me check with the Stencil team to see if I'm doing something wrong.

Comment (liamdebeasi) [reply core/src/components.d.ts]: Fixed in https://github.com/ionic-team/ionic-framework/pull/27647/commits/de54350bb6540b246df3207bb16422461dc59004

File: core/src/components/textarea/textarea.tsx (hunk 7 above)
Comment (liamdebeasi) [reply core/src/components/textarea/textarea.tsx]: See https://github.com/ionic-team/ionic-framework/pull/27647#discussion_r1230172878
